### ISSUE TITLE
[FEAT] Connect to servers on LAN

### ISSUE THREAD

**Issue by fkirchw** [aclist/dztui#84]
I am setting up a small Namalsk server on a Windows 11 machine for a small event. Because I do not have another Windows machine available I tried testing the server using my steam deck. When trying to access the server by IP using dzgui, I get the message "Failed to retrieve IP metadata". The server machine is reachable from the steam deck (ping). I configured the firewall of the windows machine to allow all traffic for the dzserver executable. I can connect fine to other community servers using the dzgui/steam deck combo.
Edit: The official launcher does find the server on steam deck

**Comment by aclist** [comment]
Hello, thanks for the message. This is related to #66 and was impossible until the recent release of version 4.0.0. While not fully implemented, we have the basic foundation in place now for making this doable (I believe), which was part of the motivation for the changes made in 4.0.0. 

Given some time and your help, I could work on getting this feature in for the next release.

**Comment by aclist** [comment]
Thank you. Instructions follow:

0. Launch DZGUI from the terminal.
1. Navigate to Advanced Options and select Toggle Branch. This will put you on the Testing branch and download that version. Exit and restart the application.
2. Use the `Connect by IP` or `Add by IP` options and provide the IP and listening port of the local server. A preliminary but non-final implementation of local IP validation is in place. This step may work end-to-end, but is likely to partially fail; check the terminal output for meaningful errors.
3. Invoke the command `python3 $HOME/.local/share/dzgui/helpers/query.py LANSERVERIP LANSERVERPORT info`, where `LANSERVERIP` and `LANSERVERPORT` are the respective IP and listening port of the local server.
4. Similar to the above, invoke the command `python3 $HOME/.local/share/dzgui/helpers/query.py LANSERVERIP LANSERVERPORT rules`, where `LANSERVERIP` and `LANSERVERPORT` are the respective IP and listening port of the local server.

Attach any meaningful output from steps 2-4 above.

You can revert the branch toggle to return to the Stable branch by repeating step 1 above.

**Comment by fkirchw** [comment]
Step 3 fails with message: Failed to retrieve metadata. There is no terminal output. Contents of DZGUI_DEBUG.log:
```[2023-12-06 15:55:38] [INFO] run_depcheck
[2023-12-06 15:55:38] [INFO] watcher_deps
[2023-12-06 15:55:38] [INFO] check_architecture
[2023-12-06 15:55:38] [INFO] Setting architecture to 'Steam Deck'
[2023-12-06 15:55:38] [INFO] check_version
[2023-12-06 15:55:38] [INFO] check_branch
[2023-12-06 15:55:38] [INFO] Branch is testing
[2023-12-06 15:55:39] [INFO] Upstream version is 4.1.0.rc-4
[2023-12-06 15:55:39] [INFO] Local version is same as upstream
[2023-12-06 15:55:39] [INFO] check_unmerged
[2023-12-06 15:55:39] [INFO] check_map_count
[2023-12-06 15:55:39] [INFO] fetch_helpers
[2023-12-06 15:55:40] [INFO] config
[2023-12-06 15:55:40] [INFO] steam_deps
[2023-12-06 15:55:40] [INFO] steam_cmd set to steam
[2023-12-06 15:55:40] [INFO] run_varcheck
[2023-12-06 15:55:40] [INFO] stale_symlinks
[2023-12-06 15:55:40] [INFO] setup
[2023-12-06 15:55:40] [INFO] check_news
[2023-12-06 15:55:40] [INFO] News: NEWS: Major 4.0.0 update. Full details via the View Changelog menu option.
[2023-12-06 15:55:40] [INFO] main_menu
[2023-12-06 15:55:40] [INFO] Setting mode
[2023-12-06 15:55:40] [INFO] set_mode
[2023-12-06 15:55:40] [INFO] Mode is normal
[2023-12-06 15:55:40] [INFO] set_header
[2023-12-06 15:55:40] [INFO] Header mode is main_menu
[2023-12-06 15:55:40] [INFO] print_news
[2023-12-06 15:55:42] [INFO] set_header rc is 0
[2023-12-06 15:55:56] [INFO] Local LAN response follows
[2023-12-06 15:55:56] [INFO] 
[2023-12-06 15:55:56] [WARN] Failed to retrieve server metadata. Check IP:PORT combination and try again.
[2023-12-06 15:55:58] [INFO] set_header
[2023-12-06 15:55:58] [INFO] Header mode is main_menu
[2023-12-06 15:55:58] [INFO] print_news
[2023-12-06 15:56:02] [INFO] set_header rc is 1
[2023-12-06 15:56:02] [INFO] Returning from main menu
```

neither 3 or 4 produce output.

**Comment by fkirchw** [comment]
No, I initially used the game port, where the timeout occurred. I then used the query port which supplied the expected output. Supplying the query port for step 2 gives the same Metadata warning window as in my first post.

**Comment by aclist** [comment]
OK, understood. I called it the listening port because I thought this would clarify that I meant the query port, but maybe that was vague. 

I've made a few changes to the control flow in version `4.1.0.rc-5` and increased the verbosity of the logging. You should supply the LAN server IP and query port in the format `127.0.0.1:27016` when prompted to in the input field and it should use A2S accordingly and at least give some more insight in the logs. We are doing black box development here a bit, but the solution should come soon enough.

I am currently rewriting the entire GUI framework to decouple the business logic from the menus and logging, which should make debugging better in the future, but right now what we have is a bit of a bricolage of moving parts.

**Comment by aclist** [comment]
I found another small bug in the control flow that was blocking progress even if the local server returned a well-formatted response. I've fixed that and bumped the version again to `4.1.0.rc-7`.

**Comment by aclist** [comment]
Thank you; the logs helped. The query port was not getting emitted as raw output. I've fixed this and bumped the version again. The query process should succeed now.

**Comment by fkirchw** [comment]
New version shows dialog saying: 
```
This is a dry run.
These options would have been used to launch the game:

steam -applaunch 221100 -connect=192.168.0.183:2302 -nolaunc
her -nosplash -name=USERNAME -skipintro "-mod=@024b5185;@271
90e6d"
```

Debug log output:
```
[2023-12-08 21:51:34] [INFO] run_depcheck
[2023-12-08 21:51:34] [INFO] watcher_deps
[2023-12-08 21:51:34] [INFO] check_architecture
[2023-12-08 21:51:34] [INFO] Setting architecture to 'Steam Deck'
[2023-12-08 21:51:34] [INFO] check_version
[2023-12-08 21:51:34] [INFO] check_branch
[2023-12-08 21:51:34] [INFO] Branch is testing
[2023-12-08 21:51:35] [INFO] Upstream version is 4.1.0.rc-7
[2023-12-08 21:51:35] [INFO] Local version is same as upstream
[2023-12-08 21:51:35] [INFO] check_unmerged
[2023-12-08 21:51:35] [INFO] check_map_count
[2023-12-08 21:51:35] [INFO] fetch_helpers
[2023-12-08 21:51:36] [INFO] config
[2023-12-08 21:51:36] [INFO] steam_deps
[2023-12-08 21:51:36] [INFO] steam_cmd set to steam
[2023-12-08 21:51:36] [INFO] run_varcheck
[2023-12-08 21:51:36] [INFO] stale_symlinks
[2023-12-08 21:51:36] [INFO] setup
[2023-12-08 21:51:36] [INFO] check_news
[2023-12-08 21:51:36] [INFO] News: NEWS: Major 4.0.0 update. Full details via the View Changelog menu option.
[2023-12-08 21:51:36] [INFO] main_menu
[2023-12-08 21:51:36] [INFO] Setting mode
[2023-12-08 21:51:36] [INFO] set_mode
[2023-12-08 21:51:36] [INFO] Mode is debug
[2023-12-08 21:51:36] [INFO] set_header
[2023-12-08 21:51:36] [INFO] Header mode is main_menu
[2023-12-08 21:51:36] [INFO] print_news
[2023-12-08 21:51:40] [INFO] set_header rc is 0
[2023-12-08 21:51:40] [INFO] set_mode
[2023-12-08 21:51:40] [INFO] Mode is debug
[2023-12-08 21:51:41] [INFO] set_mode
[2023-12-08 21:51:41] [INFO] Mode is normal
[2023-12-08 21:51:44] [INFO] main_menu
[2023-12-08 21:51:44] [INFO] Setting mode
[2023-12-08 21:51:44] [INFO] set_mode
[2023-12-08 21:51:44] [INFO] Mode is normal
[2023-12-08 21:51:44] [INFO] set_header
[2023-12-08 21:51:44] [INFO] Header mode is main_menu
[2023-12-08 21:51:44] [INFO] print_news
[2023-12-08 21:51:46] [INFO] set_header rc is 0
[2023-12-08 21:51:57] [INFO] Given LAN IP was 192.168.0.183
[2023-12-08 21:51:57] [INFO] Given LAN port was 27016
[2023-12-08 21:51:57] [A2S] Querying '192.168.0.183:27016' with mode 'info'
[2023-12-08 21:51:57] [INFO] {"name": "Local Dayz Server", "address": "192.168.0.183:2302", "count": "0/30", "keywords": "battleye,no3rd,external,privHive,shard,lqs0,etm2.500000,entm9.000000,mod,11:08", "stat": "online", "qport": "27016", "ping": "130ms"}
[2023-12-08 21:51:58] [INFO] Found 'Local Dayz Server' at 192.168.0.183:2302:"27016"
[2023-12-08 21:51:58] [INFO] Querying 192.168.0.183:2302:"27016"
[2023-12-08 21:51:58] [A2S] Querying '192.168.0.183:"27016"' with mode 'rules'
[2023-12-08 21:51:59] [INFO] set_header
[2023-12-08 21:51:59] [INFO] Header mode is main_menu
[2023-12-08 21:51:59] [INFO] print_news
[2023-12-08 21:52:01] [INFO] set_header rc is 1
[2023-12-08 21:52:01] [INFO] Returning from main menu

```

**Comment by fkirchw** [comment]
executing the command from the dialog successfully launches the game and joins the server, with correct mod configuration

**Comment by aclist** [comment]
If you navigate to `Advanced Options > Toggle debug mode` and turn it to `normal` (off), you should be able to connect directly to the server without the dry run.

**Comment by aclist** [comment]
Make sure you are using version `4.1.0.rc-8`, though, as some of the logs you posted look polluted.

**Comment by fkirchw** [comment]
With debug mode toggled off, the game starts and joins successfully, the output log for one such run is:
```
[2023-12-09 15:29:02] [INFO] run_depcheck
[2023-12-09 15:29:02] [INFO] watcher_deps
[2023-12-09 15:29:02] [INFO] check_architecture
[2023-12-09 15:29:02] [INFO] Setting architecture to 'Steam Deck'
[2023-12-09 15:29:02] [INFO] check_version
[2023-12-09 15:29:02] [INFO] check_branch
[2023-12-09 15:29:02] [INFO] Branch is testing
[2023-12-09 15:29:02] [INFO] Upstream version is 4.1.0.rc-8
[2023-12-09 15:29:02] [INFO] Local version is same as upstream
[2023-12-09 15:29:02] [INFO] check_unmerged
[2023-12-09 15:29:02] [INFO] check_map_count
[2023-12-09 15:29:02] [INFO] fetch_helpers
[2023-12-09 15:29:03] [INFO] config
[2023-12-09 15:29:03] [INFO] steam_deps
[2023-12-09 15:29:03] [INFO] steam_cmd set to steam
[2023-12-09 15:29:03] [INFO] run_varcheck
[2023-12-09 15:29:03] [INFO] stale_symlinks
[2023-12-09 15:29:03] [INFO] setup
[2023-12-09 15:29:03] [INFO] check_news
[2023-12-09 15:29:03] [INFO] News: NEWS: Major 4.0.0 update. Full details via the View Changelog menu option.
[2023-12-09 15:29:03] [INFO] main_menu
[2023-12-09 15:29:03] [INFO] Setting mode
[2023-12-09 15:29:03] [INFO] set_mode
[2023-12-09 15:29:03] [INFO] Mode is normal
[2023-12-09 15:29:03] [INFO] set_header
[2023-12-09 15:29:03] [INFO] Header mode is main_menu
[2023-12-09 15:29:03] [INFO] print_news
[2023-12-09 15:29:09] [INFO] set_header rc is 0
[2023-12-09 15:29:20] [A2S] Querying '192.168.0.183:27016' with mode 'info'
[2023-12-09 15:29:24] [INFO] set_header
[2023-12-09 15:29:24] [INFO] Header mode is connect
[2023-12-09 15:29:24] [INFO] print_news
[2023-12-09 15:29:27] [INFO] Querying 192.168.0.183:2302:27016
[2023-12-09 15:29:27] [A2S] Querying '192.168.0.183:27016' with mode 'rules'
[2023-12-09 15:29:28] [INFO] Server returned modlist: 2289456201 2289461232 
[2023-12-09 15:29:33] [INFO] set_header
[2023-12-09 15:29:33] [INFO] Header mode is connect
[2023-12-09 15:29:33] [INFO] print_news
```

**Comment by aclist** [comment]
Well, I would consider this a success. Do you feel anything is missing from the above feature?

**Comment by fkirchw** [comment]
No, it seems to work without any issues. Thank you for looking into this.

**Comment by fkirchw** [comment]
Standard timeout in a2s is 3 seconds i think, i have not had any benefits from increasing that.

**Comment by aclist** [comment]
Thank you for your support. This feature is now available in the Stable branch, version 4.1.0. Please submit another ticket if you encounter other issues.
